### PR TITLE
Added more support for TypeScript

### PR DESCRIPTION
**Pull request by JasonHK**
Please do not update the README or Change Log, we will do this when we merge your PR.

This PR (delete as applicable)

* Updates the Documentation
* Adds a new feature

Describe the changes below:

* Added support for `interface`s.
* Added support for function and method overloads.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.GameObjects.GameObject`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.GameObjects.Video`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Input.Gamepad.Gamepad`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Input.Gamepad.GamepadPlugin`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Input.InputManager#events`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Input.InputPlugin`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Input.Keyboard.Key`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Input.Keyboard.KeyboardPlugin`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Loader.LoaderPlugin`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Physics.Arcade.World`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Physics.Impact.World`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Physics.Matter.World`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Scene#events`.
* Added definitions for `Phaser.Events.EventEmitter` methods of `Phaser.Scenes.Systems#events`.

Potential issues:

* Because I was not able to install the dependencies of `phaser3-docs` successfully, I was not able to verify the modifications to the JSDoc will not affect the documentation.

### CODE REVIEW

**Comment by photonstorm** [comment]
Thanks for this PR. I need to ensure this change won't break the JSDocs output, so please give me some time to do a local checkout and build the docs to make sure they still generate properly.

File: src/scene/Systems.js [hunk 1]
```diff
@@ -240,7 +240,7 @@ var Systems = new Class({
          * In the default set-up you can access this from within a Scene via the `this.events` property.
          *
          * @name Phaser.Scenes.Systems#events
-         * @type {Phaser.Events.EventEmitter}
+         * @type {Phaser.Types.Scenes.SceneEventsEmitterr}
```
Comment (samme) [inline src/scene/Systems.js]: Should be `SceneEventsEmitter`

**Comment by photonstorm** [comment]
Hi - I've checked this out locally and built the docs from it. Most of the changes seem fine, but some of them have broken the docs output (or made the output not make any sense).

For example, the change to `Phaser.Display.Align` to an `enum` breaks the namespace output page. It goes from this, with all the correct consts defined on it:

image

to this, with all the consts missing:

image

When the Event Emitter has been changed, such as in `game.events` it replaces the output with this:

image

Which isn't really the correct type, but I don't mind too much. However, clicking it takes you to this page:

image

All of the source links are wrong. This happens through-out the whole docs where Source links now point to the typedef line instead of the actual source code.

For some reason, nearly all Event Emitter methods are now listed twice on any class that inherits the event emitter (you can see it in the screenshot above, but it also appears in the GameObject class, etc).

As it stands, I cannot merge this PR, sorry. I would urge you to try and get into a position where you can build the jsdocs locally. As important as TypeScript is, it's not worth compromising the docs for I'm afraid. However, I'm sure there is a solution. Perhaps you ought to create a custom `@typescript` tag to hold advanced defs information that the Parser can pick-up and use? and that way the docs output would ignore it and both sides would be safe.

**Comment by photonstorm** [comment]
Closing this as it's been quite a few months and sadly it still breaks the JSDoc output. Happy to revisit if we can move past that fact!

**Pull request closed without merging**